File: problems/496.下一个更大元素-i.py
#
# @lc app=leetcode.cn id=496 lang=python3
#
# [496] 下一个更大元素 I
#

# @lc code=start
class Solution:
    def nextGreaterElement(self, nums1: list[int], nums2: list[int]) -> list[int]:
        # 先计算 num2 的每个数的下一个更大元素的值
        nextNums2 = [-1] * len(nums2)
        s = []  # 栈
        i = len(nums2) - 1
        while i >= 0:
            while 0 != len(s) and nums2[i] >= s[len(s)-1]:
                s.pop(len(s)-1)
            if 0 == len(s):
                nextNums2[i] = -1
            else:
                nextNums2[i] = s[len(s)-1]
            s.insert(len(s), nums2[i])
            i -= 1
        # 遍历 nums1 拿到对应的值即可
        ans = []
        for e in nums1:
            ans.append(nextNums2[nums2.index(e)])
        return ans


# @lc code=end

File: problems/496.下一个更大元素-i.cpp
/*
 * @lc app=leetcode.cn id=496 lang=cpp
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start
#include <iostream>
using namespace std;
#include <algorithm>
#include <stack>
#include <vector>

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        // 先处理nums2的，用单调栈去处理，然后再根据nums1去做映射就好了
        // 单调栈，从后往前，每次进入的时候判断本值与栈中元素的大小，维持栈的单调性，在本题中我们需要找到最近的更大的元素，因此如果我想要进入的元素更大，那么后面的元素一定不是答案，出栈
        stack<int> s;
        vector<int> v(nums2.size());  // 存储num2的每个值
        for (int i = nums2.size() - 1; i >= 0; --i) {
            while (false == s.empty() and s.top() <= nums2[i])
                s.pop();

            v[i] = s.empty() ? -1 : s.top();
            s.push(nums2[i]);
        }

        // 做映射
        vector<int> res;
        for (auto& val : nums1)
            res.push_back(v[find(nums2.begin(), nums2.end(), val) - nums2.begin()]);
        return res;
    }
};
// @lc code=end
